In [62]:
import pandas as pd
import requests
ad_df_url = "https://storage.googleapis.com/karmicseed-assessments-data/ads_performance.json"
response = requests.get(ad_df_url)
data_json =  response.json()
print(data_json[:2])

[{'date': '2025-09-01', 'SKU': 'MN-16', 'impressions': '2016', 'clicks': '14', 'costPerClick': '1.998', 'clickThroughRate': '0.694', 'cost': '27.97', 'spend': '27.97', 'currency-code': 'USD', 'purchases1d': '1', 'purchases7d': '1', 'purchases14d': '1', 'purchases30d': '1', 'unitsSoldClicks1d': '1', 'unitsSoldClicks7d': '1', 'unitsSoldClicks14d': '1', 'unitsSoldClicks30d': '1', 'sales1d': '18.89', 'sales7d': '18.89', 'sales14d': '18.89', 'sales30d': '18.89', 'acosClicks7d': '148.0677607', 'acosClicks14d': '148.0677607', 'roasClicks7d': '0.6753664641', 'roasClicks14d': '0.6753664641'}, {'date': '2025-09-01', 'SKU': 'MN-15', 'impressions': '1810', 'clicks': '33', 'costPerClick': '2.062', 'clickThroughRate': '1.823', 'cost': '68.06', 'spend': '68.06', 'currency-code': 'USD', 'purchases1d': '4', 'purchases7d': '6', 'purchases14d': '6', 'purchases30d': '6', 'unitsSoldClicks1d': '4', 'unitsSoldClicks7d': '6', 'unitsSoldClicks14d': '6', 'unitsSoldClicks30d': '6', 'sales1d': '107.45', 'sales7d'

In [63]:
import json
print(json.dumps(data_json[:1], indent=4))

[
    {
        "date": "2025-09-01",
        "SKU": "MN-16",
        "impressions": "2016",
        "clicks": "14",
        "costPerClick": "1.998",
        "clickThroughRate": "0.694",
        "cost": "27.97",
        "spend": "27.97",
        "currency-code": "USD",
        "purchases1d": "1",
        "purchases7d": "1",
        "purchases14d": "1",
        "purchases30d": "1",
        "unitsSoldClicks1d": "1",
        "unitsSoldClicks7d": "1",
        "unitsSoldClicks14d": "1",
        "unitsSoldClicks30d": "1",
        "sales1d": "18.89",
        "sales7d": "18.89",
        "sales14d": "18.89",
        "sales30d": "18.89",
        "acosClicks7d": "148.0677607",
        "acosClicks14d": "148.0677607",
        "roasClicks7d": "0.6753664641",
        "roasClicks14d": "0.6753664641"
    }
]


In [64]:
ad_df = pd.DataFrame(data_json)
ad_df.head()

date    SKU impressions clicks costPerClick clickThroughRate   cost  \
0  2025-09-01  MN-16        2016     14        1.998            0.694  27.97   
1  2025-09-01  MN-15        1810     33        2.062            1.823  68.06   
2  2025-09-01  MN-14         613      9        2.850            1.468  25.65   
3  2025-09-01  MN-17        1271      3        5.620            0.236  16.86   
4  2025-09-01  MN-25        1036     10        4.340            0.965   43.4   

   spend currency-code purchases1d  ... unitsSoldClicks14d unitsSoldClicks30d  \
0  27.97           USD           1  ...                  1                  1   
1  68.06           USD           4  ...                  6                  6   
2  25.65           USD           0  ...                  1                  1   
3  16.86           USD           2  ...                  9                  9   
4   43.4           USD           1  ...                  8                  8   

  sales1d sales7d sales14d sales30d acosClicks7d acosClicks14d  roasClicks7d  \
0   18.89   18.89    18.89    18.89  148.0677607   148.0677607  0.6753664641   
1  107.45  161.31   161.31   161.31  42.19205257   42.19205257   2.370114605   
2       0   90.98    90.98    90.98  28.19300945   28.19300945   3.546978558   
3  274.41  274.41   274.41   274.41  6.144090959   6.144090959   16.27580071   
4   218.3  851.24   851.24   851.24  5.098444622   5.098444622   19.61382488   

  roasClicks14d  
0  0.6753664641  
1   2.370114605  
2   3.546978558  
3   16.27580071  
4   19.61382488  

[5 rows x 25 columns]

In [65]:
ad_df.columns

Index(['date', 'SKU', 'impressions', 'clicks', 'costPerClick',
       'clickThroughRate', 'cost', 'spend', 'currency-code', 'purchases1d',
       'purchases7d', 'purchases14d', 'purchases30d', 'unitsSoldClicks1d',
       'unitsSoldClicks7d', 'unitsSoldClicks14d', 'unitsSoldClicks30d',
       'sales1d', 'sales7d', 'sales14d', 'sales30d', 'acosClicks7d',
       'acosClicks14d', 'roasClicks7d', 'roasClicks14d'],
      dtype='object')

In [66]:
ad_df.isna().sum()

date                  0
SKU                   0
impressions           0
clicks                0
costPerClick          0
clickThroughRate      0
cost                  0
spend                 0
currency-code         0
purchases1d           0
purchases7d           0
purchases14d          0
purchases30d          0
unitsSoldClicks1d     0
unitsSoldClicks7d     0
unitsSoldClicks14d    0
unitsSoldClicks30d    0
sales1d               0
sales7d               0
sales14d              0
sales30d              0
acosClicks7d          0
acosClicks14d         0
roasClicks7d          0
roasClicks14d         0
dtype: int64

In [67]:
import numpy as np
MISSING_VALUES = [
    "", " ", "  ", "NA", "N/A", "na", "null", "None", "-", "--"
]
ad_df.replace(MISSING_VALUES, np.nan, inplace=True)

In [68]:
ads_numeric_cols = [
    "impressions", "clicks", "costPerClick", "clickThroughRate",
    "cost", "spend",
    "purchases1d", "purchases7d", "purchases14d", "purchases30d",
    "unitsSoldClicks1d", "unitsSoldClicks7d",
    "unitsSoldClicks14d", "unitsSoldClicks30d",
    "sales1d", "sales7d", "sales14d", "sales30d",
    "acosClicks7d", "acosClicks14d",
    "roasClicks7d", "roasClicks14d"
]

ad_df[ads_numeric_cols] = ad_df[ads_numeric_cols].apply(
    pd.to_numeric, errors="coerce"
)


In [69]:
ad_df.isna().sum()

date                     0
SKU                      0
impressions              0
clicks                   0
costPerClick          3708
clickThroughRate        62
cost                     0
spend                    0
currency-code            0
purchases1d              0
purchases7d              0
purchases14d             0
purchases30d             0
unitsSoldClicks1d        0
unitsSoldClicks7d        0
unitsSoldClicks14d       0
unitsSoldClicks30d       0
sales1d                  0
sales7d                  0
sales14d                 0
sales30d                 0
acosClicks7d          7934
acosClicks14d         7926
roasClicks7d          3708
roasClicks14d         3708
dtype: int64

In [70]:
filtered_rows = ad_df[
    ad_df["costPerClick"].isna() & (ad_df["clicks"] > 0)
]

filtered_rows


Empty DataFrame
Columns: [date, SKU, impressions, clicks, costPerClick, clickThroughRate, cost, spend, currency-code, purchases1d, purchases7d, purchases14d, purchases30d, unitsSoldClicks1d, unitsSoldClicks7d, unitsSoldClicks14d, unitsSoldClicks30d, sales1d, sales7d, sales14d, sales30d, acosClicks7d, acosClicks14d, roasClicks7d, roasClicks14d]
Index: []

[0 rows x 25 columns]

##### The costPerClick column contains missing values (3708) for rows where the number of clicks is zero. Since cost per click is undefined when no clicks occur, these values represent periods of no ad engagement rather than data quality issues. Therefore, such missing values are safely filled with 0.

In [71]:
ad_df["costPerClick"] = ad_df["costPerClick"].fillna(0)


In [72]:
ad_df["clickThroughRate"] = ad_df["clickThroughRate"].fillna(0)


In [73]:
ad_df["roasClicks7d"] = ad_df["roasClicks7d"].fillna(0)
ad_df["roasClicks14d"] = ad_df["roasClicks14d"].fillna(0)

In [74]:
filtered_rows_acos = ad_df[
    ad_df["acosClicks7d"].isna() ]

filtered_rows_acos


date    SKU  impressions  clicks  costPerClick  clickThroughRate  \
8      2025-09-01  MN-37           23       0         0.000             0.000   
18     2025-09-01  MN-34          116       4         2.410             3.448   
19     2025-09-01  MN-35          830      12         2.275             1.446   
26     2025-09-01  MN-47           49       0         0.000             0.000   
27     2025-09-01  MN-35          232       2         1.290             0.862   
...           ...    ...          ...     ...           ...               ...   
12772  2025-11-30  MN-37           49       0         0.000             0.000   
12773  2025-11-30  MN-37           32       0         0.000             0.000   
12774  2025-11-30  MN-27           63       0         0.000             0.000   
12775  2025-11-30  MN-43          434       0         0.000             0.000   
12776  2025-11-30  MN-23         1187       6         3.455             0.505   

        cost  spend currency-code  purchases1d  ...  unitsSoldClicks14d  \
8       0.00   0.00           USD            0  ...                   0   
18      9.64   9.64           USD            0  ...                   0   
19     27.30  27.30           USD            0  ...                   0   
26      0.00   0.00           USD            0  ...                   0   
27      2.58   2.58           USD            0  ...                   0   
...      ...    ...           ...          ...  ...                 ...   
12772   0.00   0.00           USD            0  ...                   0   
12773   0.00   0.00           USD            0  ...                   0   
12774   0.00   0.00           USD            0  ...                   0   
12775   0.00   0.00           USD            0  ...                   0   
12776  20.73  20.73           USD            0  ...                   0   

       unitsSoldClicks30d  sales1d  sales7d  sales14d  sales30d  acosClicks7d  \
8                       0      0.0      0.0       0.0       0.0           NaN   
18                      0      0.0      0.0       0.0       0.0           NaN   
19                      0      0.0      0.0       0.0       0.0           NaN   
26                      0      0.0      0.0       0.0       0.0           NaN   
27                      0      0.0      0.0       0.0       0.0           NaN   
...                   ...      ...      ...       ...       ...           ...   
12772                   0      0.0      0.0       0.0       0.0           NaN   
12773                   0      0.0      0.0       0.0       0.0           NaN   
12774                   0      0.0      0.0       0.0       0.0           NaN   
12775                   0      0.0      0.0       0.0       0.0           NaN   
12776                   0      0.0      0.0       0.0       0.0           NaN   

       acosClicks14d  roasClicks7d  roasClicks14d  
8                NaN           0.0            0.0  
18               NaN           0.0            0.0  
19               NaN           0.0            0.0  
26               NaN           0.0            0.0  
27               NaN           0.0            0.0  
...              ...           ...            ...  
12772            NaN           0.0            0.0  
12773            NaN           0.0            0.0  
12774            NaN           0.0            0.0  
12775            NaN           0.0            0.0  
12776            NaN           0.0            0.0  

[7934 rows x 25 columns]

In [75]:
acos_agg = (
    ad_df
    .groupby("SKU", as_index=False)
    .agg({
        "acosClicks7d": "mean",
        "acosClicks14d": "mean"
    })
)
acos_agg

SKU  acosClicks7d  acosClicks14d
0   MN-01     31.491089      31.486923
1   MN-02     17.450712      17.429955
2   MN-03     14.699685      14.183546
3   MN-04     38.972894      40.747658
4   MN-05     28.954224      25.843374
5   MN-06     49.906491      49.906491
6   MN-07     10.698216      10.698216
7   MN-08      4.614372       4.614372
8   MN-09     13.280875      13.280875
9   MN-10     18.167006      18.167006
10  MN-11     20.120138      19.915199
11  MN-12     37.120997      37.854601
12  MN-13     21.838850      21.889256
13  MN-14     39.577595      39.266970
14  MN-15     36.639962      36.504738
15  MN-16     36.267394      36.226361
16  MN-17     33.469154      33.921212
17  MN-18     17.877862      17.877862
18  MN-19     21.304418      21.259657
19  MN-20     27.841144      27.650197
20  MN-21     20.302573      20.261531
21  MN-22     29.912885      29.912885
22  MN-23     20.919278      20.877112
23  MN-24     24.197673      24.151847
24  MN-25     24.029368      23.889000
25  MN-26     23.794540      23.756626
26  MN-27     26.476689      26.476689
27  MN-28     31.892060      31.794818
28  MN-29     30.954084      30.863715
29  MN-30     41.420546      41.399948
30  MN-31     31.393345      30.909478
31  MN-32     34.329271      34.329271
32  MN-33     20.913204      20.649876
33  MN-34     27.686315      27.686315
34  MN-35     28.026200      28.026200
35  MN-36     30.397855      30.369378
36  MN-37     20.675543      20.675543
37  MN-38     40.367727      38.642590
38  MN-39     35.948862      35.948862
39  MN-40     22.018600      22.002956
40  MN-41     11.957454      11.908474
41  MN-42     20.989055      20.984806
42  MN-43     77.141743      76.702818
43  MN-44     25.937288      25.937288
44  MN-45     15.638046      15.221962
45  MN-46     18.265842      18.265842
46  MN-47      9.893044       9.893044
47  MN-48     14.015364      14.015364
48  MN-49     13.676930      13.676930
49  MN-50     12.446630      12.421240

In [76]:
ad_df = ad_df.merge(
    acos_agg,
    on="SKU",
    how="left",
    suffixes=("", "_sku_avg")
)


In [77]:
ad_df["acosClicks14d_clean"] = ad_df["acosClicks14d"].fillna(
    ad_df["acosClicks14d_sku_avg"]
)

ad_df["acosClicks7d_clean"] = ad_df["acosClicks7d"].fillna(
    ad_df["acosClicks7d_sku_avg"]
)


In [78]:
ad_df

date    SKU  impressions  clicks  costPerClick  clickThroughRate  \
0      2025-09-01  MN-16         2016      14         1.998             0.694   
1      2025-09-01  MN-15         1810      33         2.062             1.823   
2      2025-09-01  MN-14          613       9         2.850             1.468   
3      2025-09-01  MN-17         1271       3         5.620             0.236   
4      2025-09-01  MN-25         1036      10         4.340             0.965   
...           ...    ...          ...     ...           ...               ...   
12772  2025-11-30  MN-37           49       0         0.000             0.000   
12773  2025-11-30  MN-37           32       0         0.000             0.000   
12774  2025-11-30  MN-27           63       0         0.000             0.000   
12775  2025-11-30  MN-43          434       0         0.000             0.000   
12776  2025-11-30  MN-23         1187       6         3.455             0.505   

        cost  spend currency-code  purchases1d  ...  sales14d  sales30d  \
0      27.97  27.97           USD            1  ...     18.89     18.89   
1      68.06  68.06           USD            4  ...    161.31    161.31   
2      25.65  25.65           USD            0  ...     90.98     90.98   
3      16.86  16.86           USD            2  ...    274.41    274.41   
4      43.40  43.40           USD            1  ...    851.24    851.24   
...      ...    ...           ...          ...  ...       ...       ...   
12772   0.00   0.00           USD            0  ...      0.00      0.00   
12773   0.00   0.00           USD            0  ...      0.00      0.00   
12774   0.00   0.00           USD            0  ...      0.00      0.00   
12775   0.00   0.00           USD            0  ...      0.00      0.00   
12776  20.73  20.73           USD            0  ...      0.00      0.00   

       acosClicks7d  acosClicks14d  roasClicks7d  roasClicks14d  \
0        148.067761     148.067761      0.675366       0.675366   
1         42.192053      42.192053      2.370115       2.370115   
2         28.193009      28.193009      3.546979       3.546979   
3          6.144091       6.144091     16.275801      16.275801   
4          5.098445       5.098445     19.613825      19.613825   
...             ...            ...           ...            ...   
12772           NaN            NaN      0.000000       0.000000   
12773           NaN            NaN      0.000000       0.000000   
12774           NaN            NaN      0.000000       0.000000   
12775           NaN            NaN      0.000000       0.000000   
12776           NaN            NaN      0.000000       0.000000   

       acosClicks7d_sku_avg  acosClicks14d_sku_avg  acosClicks14d_clean  \
0                 36.267394              36.226361           148.067761   
1                 36.639962              36.504738            42.192053   
2                 39.577595              39.266970            28.193009   
3                 33.469154              33.921212             6.144091   
4                 24.029368              23.889000             5.098445   
...                     ...                    ...                  ...   
12772             20.675543              20.675543            20.675543   
12773             20.675543              20.675543            20.675543   
12774             26.476689              26.476689            26.476689   
12775             77.141743              76.702818            76.702818   
12776             20.919278              20.877112            20.877112   

       acosClicks7d_clean  
0              148.067761  
1               42.192053  
2               28.193009  
3                6.144091  
4                5.098445  
...                   ...  
12772           20.675543  
12773           20.675543  
12774           26.476689  
12775           77.141743  
12776           20.919278  

[12777 rows x 29 columns]

In [79]:
ad_df.drop(["acosClicks7d","acosClicks14d"],axis = "columns",inplace = True)

In [80]:
ad_df.isna().sum()

date                     0
SKU                      0
impressions              0
clicks                   0
costPerClick             0
clickThroughRate         0
cost                     0
spend                    0
currency-code            0
purchases1d              0
purchases7d              0
purchases14d             0
purchases30d             0
unitsSoldClicks1d        0
unitsSoldClicks7d        0
unitsSoldClicks14d       0
unitsSoldClicks30d       0
sales1d                  0
sales7d                  0
sales14d                 0
sales30d                 0
roasClicks7d             0
roasClicks14d            0
acosClicks7d_sku_avg     0
acosClicks14d_sku_avg    0
acosClicks14d_clean      0
acosClicks7d_clean       0
dtype: int64

In [81]:
ads_agg_df = (
    ad_df
    .groupby("SKU", as_index=False)
    .agg({
        "impressions": "sum",
        "clicks": "sum",
        "cost": "sum",
        "spend": "sum",
        "sales14d": "sum",
        "sales30d": "sum",
        "acosClicks14d_clean": "mean",
        "roasClicks14d": "mean"
    })
    .round(2)
)


In [82]:
ads_agg_df

SKU  impressions  clicks      cost     spend   sales14d   sales30d  \
0   MN-01       131593     985   2762.06   2762.06   14258.43   14258.43   
1   MN-02       226158    1298   2286.92   2286.92   11664.02   11664.02   
2   MN-03        74652     394    670.20    670.20    3675.34    3675.34   
3   MN-04        52363     231    903.79    903.79    2090.28    2090.28   
4   MN-05        39306     390    637.85    637.85    1779.54    1779.54   
5   MN-06        98522     428    968.46    968.46    1880.96    1880.96   
6   MN-07       219364    1380   1649.76   1649.76    9004.02    9004.02   
7   MN-08        41005     199    194.11    194.11    1444.07    1444.07   
8   MN-09        83058     439    398.62    398.62    1412.58    1412.58   
9   MN-10       122210     690   1142.65   1142.65    3648.44    3648.44   
10  MN-11       278599    1405   1698.77   1698.77    8529.35    8529.35   
11  MN-12       149641     959   1436.43   1436.43    4465.83    4465.83   
12  MN-13       340607    2783   2782.79   2782.79   11538.40   11538.40   
13  MN-14       204049    2023   5935.26   5935.26   18551.12   18551.12   
14  MN-15       446663    5623  11204.57  11204.57   35319.03   35319.03   
15  MN-16       200933    1338   2293.56   2293.56    7953.03    7953.03   
16  MN-17       259897    1133   3852.35   3852.35   10507.68   10507.68   
17  MN-18       367799    2123   3904.32   3904.32   18094.62   18094.62   
18  MN-19       149300    1280   2589.76   2589.76   16040.25   16040.25   
19  MN-20       554777    6413   3297.92   3297.92   20002.39   20002.39   
20  MN-21       596126    2909   2934.51   2934.51   15743.93   15743.93   
21  MN-22        81210     857   1646.95   1646.95    6086.37    6086.37   
22  MN-23       124212    1358   1238.32   1238.32    6798.04    6798.04   
23  MN-24       564036    3480   3432.33   3432.33   13628.76   13628.76   
24  MN-25       776060    6396  28107.19  28107.19  138640.93  138640.93   
25  MN-26       275521    1617   8107.79   8107.79   48132.26   48132.26   
26  MN-27       126221     826   4171.37   4171.37   20260.19   20260.19   
27  MN-28       374026    2803  14448.66  14448.66   63225.99   63225.99   
28  MN-29       507797    2843  10098.84  10098.84   43907.88   43907.88   
29  MN-30       193340    1444   4762.27   4762.27   18251.23   18251.23   
30  MN-31       207611     689   1670.43   1670.43    6203.92    6203.92   
31  MN-32       167861     753   2911.52   2911.52   17177.90   17177.90   
32  MN-33       120014     756   3111.18   3111.18   19121.04   19121.04   
33  MN-34        65691     401   2133.58   2133.58   10984.63   10984.63   
34  MN-35        86597     623   1901.14   1901.14    9131.41    9131.41   
35  MN-36       361721    3372   7892.54   7892.54   27207.54   27207.54   
36  MN-37        91210     457   1071.89   1071.89    3445.40    3445.40   
37  MN-38       257618    2281   4601.90   4601.90   18181.30   18181.30   
38  MN-39        44556     212    721.35    721.35    2289.02    2289.02   
39  MN-40        51489     474   1500.84   1500.84   12200.76   12200.76   
40  MN-41       158875     781   5004.61   5004.61   58209.00   58209.00   
41  MN-42       160537     874   3984.18   3984.18   33095.41   33095.41   
42  MN-43       127464     608   1608.75   1608.75    2661.62    2661.62   
43  MN-44        47541     225    342.08    342.08    1305.42    1305.42   
44  MN-45        51945     202    603.67    603.67    7757.08    7757.08   
45  MN-46        41213     181    356.70    356.70    1535.20    1535.20   
46  MN-47        15387     115    430.26    430.26    4029.08    4029.08   
47  MN-48        26507     129    711.91    711.91    5943.72    5943.72   
48  MN-49       244778    1450   1430.45   1430.45    6316.95    6316.95   
49  MN-50       190128    1263   1349.11   1349.11    8494.10    8494.10   

    acosClicks14d_clean  roasClicks14d  
0                 31.49           3.96  
1                 17.43           8.40  
2     

In [83]:
ads_agg_df["click_rate"] = np.where(
    ads_agg_df["impressions"] > 0,
    ads_agg_df["clicks"] / ads_agg_df["impressions"],
    0
)


In [84]:
ads_agg_df["cost_per_ad_sale"] = np.where(
    ads_agg_df["sales14d"] > 0,
    ads_agg_df["cost"] / ads_agg_df["sales14d"],
    np.nan
)

In [85]:
def ads_efficiency(row):
    if row["roasClicks14d"] >= 4:
        return "Efficient Ads"
    elif 2 <= row["roasClicks14d"] < 4:
        return "Moderate Ads"
    elif row["roasClicks14d"] < 2 and row["sales14d"] > 0:
        return "Inefficient Ads"
    else:
        return "No Ad Sales"

ads_agg_df["ads_efficiency"] = ads_agg_df.apply(
    ads_efficiency, axis="columns"
)


In [86]:
ads_agg_df["ad_dependency_flag"] = np.where(
    (ads_agg_df["sales14d"] > 0) &
    (ads_agg_df["roasClicks14d"] < 2.5),
    "High Ad Dependency",
    "Normal"
)


In [87]:
def pricing_strategy_ads(row):
    if row["ads_efficiency"] == "Efficient Ads":
        return "Allow Price Increase"

    if row["ads_efficiency"] == "Moderate Ads":
        return "Hold Price"

    if row["ads_efficiency"] == "Inefficient Ads":
        return "Reduce Price or Improve Listing"

    if row["ad_dependency_flag"] == "High Ad Dependency":
        return "Freeze Price Increase"

    return "Hold Price"

ads_agg_df["pricing_strategy_ads"] = ads_agg_df.apply(
    pricing_strategy_ads, axis="columns"
)


In [88]:
ads_agg_df

SKU  impressions  clicks      cost     spend   sales14d   sales30d  \
0   MN-01       131593     985   2762.06   2762.06   14258.43   14258.43   
1   MN-02       226158    1298   2286.92   2286.92   11664.02   11664.02   
2   MN-03        74652     394    670.20    670.20    3675.34    3675.34   
3   MN-04        52363     231    903.79    903.79    2090.28    2090.28   
4   MN-05        39306     390    637.85    637.85    1779.54    1779.54   
5   MN-06        98522     428    968.46    968.46    1880.96    1880.96   
6   MN-07       219364    1380   1649.76   1649.76    9004.02    9004.02   
7   MN-08        41005     199    194.11    194.11    1444.07    1444.07   
8   MN-09        83058     439    398.62    398.62    1412.58    1412.58   
9   MN-10       122210     690   1142.65   1142.65    3648.44    3648.44   
10  MN-11       278599    1405   1698.77   1698.77    8529.35    8529.35   
11  MN-12       149641     959   1436.43   1436.43    4465.83    4465.83   
12  MN-13       340607    2783   2782.79   2782.79   11538.40   11538.40   
13  MN-14       204049    2023   5935.26   5935.26   18551.12   18551.12   
14  MN-15       446663    5623  11204.57  11204.57   35319.03   35319.03   
15  MN-16       200933    1338   2293.56   2293.56    7953.03    7953.03   
16  MN-17       259897    1133   3852.35   3852.35   10507.68   10507.68   
17  MN-18       367799    2123   3904.32   3904.32   18094.62   18094.62   
18  MN-19       149300    1280   2589.76   2589.76   16040.25   16040.25   
19  MN-20       554777    6413   3297.92   3297.92   20002.39   20002.39   
20  MN-21       596126    2909   2934.51   2934.51   15743.93   15743.93   
21  MN-22        81210     857   1646.95   1646.95    6086.37    6086.37   
22  MN-23       124212    1358   1238.32   1238.32    6798.04    6798.04   
23  MN-24       564036    3480   3432.33   3432.33   13628.76   13628.76   
24  MN-25       776060    6396  28107.19  28107.19  138640.93  138640.93   
25  MN-26       275521    1617   8107.79   8107.79   48132.26   48132.26   
26  MN-27       126221     826   4171.37   4171.37   20260.19   20260.19   
27  MN-28       374026    2803  14448.66  14448.66   63225.99   63225.99   
28  MN-29       507797    2843  10098.84  10098.84   43907.88   43907.88   
29  MN-30       193340    1444   4762.27   4762.27   18251.23   18251.23   
30  MN-31       207611     689   1670.43   1670.43    6203.92    6203.92   
31  MN-32       167861     753   2911.52   2911.52   17177.90   17177.90   
32  MN-33       120014     756   3111.18   3111.18   19121.04   19121.04   
33  MN-34        65691     401   2133.58   2133.58   10984.63   10984.63   
34  MN-35        86597     623   1901.14   1901.14    9131.41    9131.41   
35  MN-36       361721    3372   7892.54   7892.54   27207.54   27207.54   
36  MN-37        91210     457   1071.89   1071.89    3445.40    3445.40   
37  MN-38       257618    2281   4601.90   4601.90   18181.30   18181.30   
38  MN-39        44556     212    721.35    721.35    2289.02    2289.02   
39  MN-40        51489     474   1500.84   1500.84   12200.76   12200.76   
40  MN-41       158875     781   5004.61   5004.61   58209.00   58209.00   
41  MN-42       160537     874   3984.18   3984.18   33095.41   33095.41   
42  MN-43       127464     608   1608.75   1608.75    2661.62    2661.62   
43  MN-44        47541     225    342.08    342.08    1305.42    1305.42   
44  MN-45        51945     202    603.67    603.67    7757.08    7757.08   
45  MN-46        41213     181    356.70    356.70    1535.20    1535.20   
46  MN-47        15387     115    430.26    430.26    4029.08    4029.08   
47  MN-48        26507     129    711.91    711.91    5943.72    5943.72   
48  MN-49       244778    1450   1430.45   1430.45    6316.95    6316.95   
49  MN-50       190128    1263   1349.11   1349.11    8494.10    8494.10   

    acosClicks14d_clean  roasClicks14d  click_rate  cost_per_ad_sale  \
0                 31.49           3.96    0.007485       

In [89]:
ads_agg_df.to_csv("Aggregrated_ads_Data.csv",index =False)

# Ads Performance Processing & Insights

#### The advertising performance data was aggregated at the SKU level to align with pricing decision granularity. Impressions, clicks, ad spend, and ad-attributed sales (14-day and 30-day windows) were summed to capture total exposure, engagement, and revenue contribution from advertising, while efficiency metrics such as 14-day ACOS and ROAS were averaged to reflect typical ad effectiveness rather than short-term spikes. 
#### From this aggregated dataset, several derived metrics were created to assess demand quality and pricing risk. 
#### A click rate metric was calculated as the ratio of clicks to impressions to measure how effectively ad visibility translated into customer interest. 
#### Cost per ad sale was computed by dividing total ad spend by 14-day ad-attributed sales, providing a direct view of how much advertising cost is required to generate one unit of ad-driven revenue. 
#### SKUs were then classified into ad efficiency categories based on 14-day ROAS thresholds, distinguishing products with efficient, moderate, inefficient, or no ad-supported demand. This classification enables clear separation between SKUs where ads confirm strong value perception and those where sales are heavily subsidized by spend. 
#### An ad dependency flag was further introduced to identify SKUs that rely on advertising to generate sales while delivering weak returns, signaling elevated pricing risk.
#### Finally, a pricing strategy recommendation was assigned based on ad efficiency and dependency, explicitly defining whether price increases should be allowed, held, reduced, or frozen. This ensures that pricing decisions are guided by the sustainability of demand rather than inflated by paid traffic.

# Business Summary
#### Advertising performance was used as a demand-quality filter within the pricing framework to ensure price increases are supported by efficient, sustainable demand. SKUs with strong ROAS and low dependency on ad spend were identified as candidates for price increases, while SKUs with inefficient or ad-dependent demand were restricted from upward pricing to prevent margin erosion. By translating raw ad metrics into clear efficiency and risk signals, this approach prevents paid traffic from masking pricing weaknesses and ensures that pricing decisions are grounded in long-term profitability rather than short-term volume.